client/meta_storage_client.go

```diff
@@ -0,0 +1,223 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pd
+
+import (
+	"context"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/meta_storagepb"
+	"github.com/pingcap/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/tikv/pd/client/grpcutil"
+	"go.uber.org/zap"
+)
+
+// MetaStorageClient is the interface for meta storage client.
+type MetaStorageClient interface {
+	// Watch watches on a key or prefix.
+	Watch(ctx context.Context, key []byte, opts ...OpOption) (chan []*meta_storagepb.Event, error)
+	// Get gets the value for a key.
+	Get(ctx context.Context, key []byte, opts ...OpOption) (*meta_storagepb.GetResponse, error)
+	// Put puts a key-value pair into meta storage.
+	Put(ctx context.Context, key []byte, value []byte, opts ...OpOption) (*meta_storagepb.PutResponse, error)
+}
+
+// metaStorageClient gets the meta storage client from current PD leader.
+func (c *client) metaStorageClient() meta_storagepb.MetaStorageClient {
+	if client := c.bc.GetServingEndpointClientConn(); client != nil {
+		return meta_storagepb.NewMetaStorageClient(client)
+	}
+	return nil
+}
+
+// Op represents available options when using meta storage client.
+type Op struct {
+	rangeEnd         []byte
+	revision         int64
+	prevKv           bool
+	lease            int64
+	limit            int64
+	isOptsWithPrefix bool
+}
+
+// OpOption configures etcd Op.
+type OpOption func(*Op)
+
+// WithLimit specifies the limit of the key.
+func WithLimit(limit int64) OpOption {
+	return func(op *Op) { op.limit = limit }
+}
+
+// WithRangeEnd specifies the range end of the key.
+func WithRangeEnd(rangeEnd []byte) OpOption {
+	return func(op *Op) { op.rangeEnd = rangeEnd }
+}
+
+// WithRev specifies the start revision of the key.
+func WithRev(revision int64) OpOption {
+	return func(op *Op) { op.revision = revision }
+}
+
+// WithPrevKV specifies the previous key-value pair of the key.
+func WithPrevKV() OpOption {
+	return func(op *Op) { op.prevKv = true }
+}
+
+// WithLease specifies the lease of the key.
+func WithLease(lease int64) OpOption {
+	return func(op *Op) { op.lease = lease }
+}
+
+// WithPrefix specifies the prefix of the key.
+func WithPrefix() OpOption {
+	return func(op *Op) {
+		op.isOptsWithPrefix = true
+	}
+}
+
+// See https://github.com/etcd-io/etcd/blob/da4bf0f76fb708e0b57763edb46ba523447b9510/client/v3/op.go#L372-L385
+func getPrefix(key []byte) []byte {
+	end := make([]byte, len(key))
+	copy(end, key)
+	for i := len(end) - 1; i >= 0; i-- {
+		if end[i] < 0xff {
+			end[i]++
+			end = end[:i+1]
+			return end
+		}
+	}
+	return []byte{0}
+}
+
+func (c *client) Put(ctx context.Context, key, value []byte, opts ...OpOption) (*meta_storagepb.PutResponse, error) {
```

Comment:
I think the implement's code should move to client.go not belongs here.

Reply:
It's the same as the keyspace client and resource manager client. I think it's more clear.